lib/internal/fs/glob.js

```diff
@@ -376,6 +396,238 @@ class Glob {
       }
     }
   }
+
+  async *#addSubpatterns(path, pattern) {
+    const seen = this.#cache.add(path, pattern);
+    if (seen) {
+      return;
+    }
+    const fullpath = resolve(this.#root, path);
+    const stat = await this.#cache.stat(fullpath);
+    const last = pattern.last;
+    const isDirectory = stat?.isDirectory() || (stat?.isSymbolicLink() && pattern.hasSeenSymlinks);
+    const isLast = pattern.isLast(isDirectory);
+    const isFirst = pattern.isFirst();
+
+    if (isFirst && isWindows && typeof pattern.at(0) === 'string' && StringPrototypeEndsWith(pattern.at(0), ':')) {
+      // Absolute path, go to root
+      this.#addSubpattern(`${pattern.at(0)}\\`, pattern.child(new SafeSet([1])));
+      return;
+    }
+    if (isFirst && pattern.at(0) === '') {
```

Comment:
Any idea how can I not duplicate it? this has `await` in it so we must be async but it's everywhere

there is a LOT of logic in here so I really don't want to duplicate it even though there are tests

Reply:
for starters - we can extract the loop under `for (let i = 0; i < children.length; i++) {` to its own method
other synchronous parts can do the same

Reply:
thanks, done